src/Test/Adapter/ReadTestCase.php

```
@@ -0,0 +1,197 @@
+        $this->assertInternalType('array', $result);
+        $this->assertArrayHasKey('contents', $result);
+        $this->assertEquals(static::FILE_CONTENT, $result['contents']);
+    }
+
+    public function testReadStream()
+    {
+        $result = $this->adapter->readStream('file.txt');
+        $this->assertPath('file.txt', $result);
+        $this->assertArrayHasKey('stream', $result);
+        $this->assertInternalType('resource', $result['stream']);
+        $this->assertEquals(static::FILE_CONTENT, fread($result['stream'], strlen(static::FILE_CONTENT)));
+        fclose($result['stream']);
+    }
+
+    public function testListingNonexistingDirectory()
+    {
+        $result = $this->adapter->listContents('nonexisting/directory');
+        $this->assertEquals([], $result);
+    }
+
+    public function testListContents()
+    {
+        $contents = $this->adapter->listContents('0', false);
+        $this->assertCount(2, $contents);
+        $this->assertArrayHasKey('type', $contents[0]);
+        $this->assertArrayHasKey('type', $contents[1]);
+    }
+
+    public function testGetMetadata()
+    {
+        $result = $this->adapter->getMetadata('file.txt');
+        $this->assertType('file', $result);
+        $this->assertPath('file.txt', $result);
+        $this->assertSize(strlen(static::FILE_CONTENT), $result);
+        $this->assertHasTimestamp($result);
+        // no mimetype in default metadata set
+    }
+
+    public function testGetSize()
+    {
+        $result = $this->adapter->getSize('file.txt');
+        $this->assertSize(strlen(static::FILE_CONTENT), $result);
+    }
+
+    public function testGetTimestamp()
+    {
+        $result = $this->adapter->getTimestamp('file.txt');
+        $this->assertHasTimestamp($result);
+    }
+
+    public function testGetMimetype()
+    {
+        $result = $this->adapter->getMimetype('file.txt');
+        $this->assertMimeType('text/plain', $result);
+    }
+}
```

When not supported it should throw an exception, same for setVisibility.